udf/add_monthly_searches.sql

```diff
@@ -0,0 +1,129 @@
+                           curr ARRAY<STRUCT<key STRING, value STRUCT<total_searches ARRAY<INT64>, tagged_searches ARRAY<INT64>, search_with_ads ARRAY<INT64>, ad_click ARRAY<INT64>>>>,
+                           submission_date DATE) AS ((
+  WITH prev_tbl AS (
+    SELECT *
+    FROM UNNEST(prev) AS p
+  )
+
+  SELECT ARRAY_AGG(
+    STRUCT(
+      key,
+      udf_add_monthly_engine_searches(
+        COALESCE(p.value, udf_engine_searches_struct()),
+        COALESCE(c.value, udf_engine_searches_struct()),
+        submission_date) AS value
+    ))
+  FROM
+      UNNEST(curr) AS c
+  FULL OUTER JOIN
+      prev_tbl AS p
+      USING (key)
+
+));
+
+-- Tests
+/*
+NOTE: These tests are structured in this way to distribute them.
+      If they were done as one row, they all run on a single
+      node and take >10 minutes to complete.
+*/
+
+WITH previous_examples AS (
+    SELECT
+        ARRAY [
+          STRUCT(
+            "google" AS key,
+            STRUCT(generate_array(11, 0, -1) AS total_searches,
+                   generate_array(12, 1, -1) AS tagged_searches,
+                   udf_zeroed_array(12) AS search_with_ads,
+                   udf_zeroed_array(12) AS ad_click) AS value
+        )] AS prev, "google" as type
+    UNION ALL
+    SELECT NULL AS prev, "null" as type
+), current_examples AS (
+    SELECT
+        ARRAY [
+          STRUCT(
+            "google" AS key,
+            STRUCT(udf_array_drop_first_and_append(udf_zeroed_array(12), 5) AS total_searches,
+                   udf_array_drop_first_and_append(udf_zeroed_array(12), 10) AS tagged_searches,
+                   udf_array_drop_first_and_append(udf_zeroed_array(12), 15) AS search_with_ads,
+                   udf_array_drop_first_and_append(udf_zeroed_array(12), 20) AS ad_click) AS value
+        )] AS curr, "google" AS type
+    UNION ALL
+    SELECT ARRAY[STRUCT(
+                    "bing" AS key,
+                    STRUCT(udf_array_drop_first_and_append(udf_zeroed_array(12), 1) AS total_searches,
+                           udf_array_drop_first_and_append(udf_zeroed_array(12), 2) AS tagged_searches,
+                           udf_array_drop_first_and_append(udf_zeroed_array(12), 3) AS search_with_ads,
+                           udf_array_drop_first_and_append(udf_zeroed_array(12), 4) AS ad_click) AS value
+)] AS curr,
+        "bing" as type
+    UNION ALL
+    SELECT NULL as curr, "null" as type
+), dates AS (
+    SELECT d AS date
+    FROM UNNEST(ARRAY ["2019-10-01", "2019-10-02"]) AS d
+), results AS (
+    SELECT udf_add_monthly_searches(p.prev, c.curr, CAST(d.date AS date)) AS res,
+      p.type AS p_type, c.type AS c_type, d.date AS date
+    FROM previous_examples p
+    CROSS JOIN current_examples c
+    CROSS JOIN dates d
+), expected AS (
+    SELECT *
+    FROM UNNEST(ARRAY [
```

Comment:
for consistency with existing code, please remove the leading `ARRAY` in front of `[` throughout this PR, unless it's being used to specify a schema for elements.